Further reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebKit/

cdumez opened the pull request:
#### fb3e81759d213897ec6e7a3dd80a04c30b56f872
<pre>
Further reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebKit/
<a href="https://bugs.webkit.org/show_bug.cgi?id=282221">https://bugs.webkit.org/show_bug.cgi?id=282221</a>

Reviewed by Geoffrey Garen and Darin Adler.

* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_names_header):
(generate_message_names_implementation):
* Source/WebKit/Shared/API/APIData.h:
(API::Data::create):
* Source/WebKit/Shared/API/Cocoa/WKRemoteObjectCoder.mm:
(-[WKRemoteObjectEncoder encodeBytes:length:forKey:]):
* Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectInterface.mm:
(initializeMethod):
(initializeMethods):
* Source/WebKit/Shared/API/c/WKArray.cpp:
(WKArrayCreate):
(WKArrayCreateAdoptingValues):
(WKArrayGetSize):
* Source/WebKit/Shared/API/c/WKData.cpp:
(WKDataCreate):
(WKDataGetSize):
* Source/WebKit/Shared/API/c/WKDictionary.cpp:
(WKDictionaryCreate):
(WKDictionaryCopyKeys):
* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringCreateWithUTF8CStringWithLength):
(WKStringGetUTF8CStringImpl):
(WKStringCopyJSString):
* Source/WebKit/Shared/API/c/WKURLRequest.cpp:
(WKURLRequestCopySettingHTTPBody):
(WKURLRequestSetDefaultTimeoutInterval):
* Source/WebKit/Shared/Cocoa/CoreIPCCFCharacterSet.h:
(WebKit::CoreIPCCFCharacterSet::dataReference const):
* Source/WebKit/Shared/Cocoa/CoreIPCDateComponents.mm:
* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtension::createHandlesForMachLookup):
* Source/WebKit/Shared/Cocoa/SandboxInitialiationParametersCocoa.mm:
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::tryApplyCachedSandbox):
* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
(WebKit::HistoryEntryDataDecoder::alignedBufferIsLargeEnoughToContain const):

Canonical link: <a href="https://commits.webkit.org/285917@main">https://commits.webkit.org/285917@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a97a0181fe1aecd588fcc68e6418225467ff30a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74138 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53567 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26949 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78511 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25376 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76255 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62700 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1352 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58295 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16633 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77205 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48445 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63775 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38705 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/73646 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45354 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21271 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23709 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/67275 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66826 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21617 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80031 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73396 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1455 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/809 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66596 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1599 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63792 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65869 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9795 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7955 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/95177 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11454 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1419 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4207 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20906 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1448 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1436 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1455 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/Shared/API/APIData.h:
```diff
@@ -49,14 +48,14 @@ class Data : public ObjectImpl<API::Object::Type::Data> {
-            copiedBytes = static_cast<uint8_t*>(fastMalloc(bytes.size()));
-            memcpy(copiedBytes, bytes.data(), bytes.size());
+        if (!bytes.empty()) {
+            copiedBytes = unsafeForgeSpan(static_cast<uint8_t*>(fastMalloc(bytes.size())), bytes.size());
```
I bet we can make a MallocSpan type, or just a fastMallocSpan helper that returns a span.

cdumez replied:
Ok, in my next PR, I will look into introducing some sort of "owning" span which calls a deleter in its destructor.

cdumez replied:
Introducing MallocSpan in https://github.com/WebKit/WebKit/pull/35950. Then I'll follow up to do more adoption if approved.

geoffreygaren commented on Source/WebKit/Shared/API/APIData.h:
```diff
@@ -71,7 +70,7 @@ class Data : public ObjectImpl<API::Object::Type::Data> {
         auto bufferSize = buffer.size();
         auto bufferPointer = buffer.releaseBuffer().leakPtr();
-        return createWithoutCopying({ bufferPointer, bufferSize }, [] (uint8_t* bytes, const void*) {
+        return createWithoutCopying(unsafeForgeSpan(bufferPointer, bufferSize), [] (uint8_t* bytes, const void*) {
```
Should we give Vector a releaseSpan() function?

darinadler replied:
Instead we could consider changing `MallocPtr` to be a `span` with ownership instead of a pointer with ownership, perhaps changing its name. Then `releaseBuffer` would always produce a `span`.

cdumez replied:
Yes, I will look into making this factoring in my next PR.

geoffreygaren commented on Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectInterface.mm:
```diff
@@ -123,10 +121,11 @@ static void initializeMethod(MethodInfo& methodInfo, Protocol *protocol, SEL sel
 {
     unsigned methodCount;
-    struct objc_method_description *methods = protocol_copyMethodDescriptionList(protocol, requiredMethods, true, &methodCount);
+    struct objc_method_description *rawMethods = protocol_copyMethodDescriptionList(protocol, requiredMethods, true, &methodCount);
```
Should we add a protocol_copyMethodDescriptionListSpan helper?

darinadler replied:
Love that suggestion. Even if it was only local here in this file I think it would be cleaner since it would not be mixed in with any other logic. But also we’d want to use an object that acts as smart pointer or smart owning `span` rather than explicitly calling `free`.

cdumez replied:
Yes, I will make this change in my next PR, once we have a "MallocSpan".

darinadler replied:
Just noticing that this will need a variant that does `free` instead of `fastFree`. Might end up wanting something with a template argument for the deleter like `unique_ptr`.

cdumez replied:
Yes, MallocPtr supports this already. So my future MallocSpan would too.

cdumez replied:
Following up in https://github.com/WebKit/WebKit/pull/36085 to address this.

geoffreygaren reviewed (approved):
r=me

darinadler commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -1615,21 +1611,21 @@ def generate_message_names_implementation(receivers):
     result.append('namespace IPC::Detail {\n')
     result.append('\n')
-    result.append('const MessageDescription messageDescriptions[static_cast<size_t>(MessageName::Count) + 1] = {\n')
+    result.append('const std::array<MessageDescription, static_cast<size_t>(MessageName::Count) + 1> messageDescriptions {\n')
```
I’d think the size could be omitted here and the compiler would deduce it from the size of the initializer list. But maybe not?

cdumez replied:
Then we'd have to omit the first template parameter too. I think it's be OK here though given the code below, it would figure out the type properly. I can likely switch to just `std::array`

darinadler commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -1571,7 +1569,7 @@ def generate_message_names_header(receivers):
     result.append('};\n')
     result.append('\n')
-    result.append('extern const MessageDescription messageDescriptions[static_cast<size_t>(MessageName::Count) + 1];\n')
+    result.append('extern const std::array<MessageDescription, static_cast<size_t>(MessageName::Count) + 1> messageDescriptions;\n')
```
Should append a blank line after this before appending the closing brace.

Could consider a `using` to name the array type so we don’t have to repeat it when defining the array in the `.cpp` file.

darinadler commented on Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectInterface.mm:
```diff
@@ -89,15 +87,15 @@ static void initializeMethod(MethodInfo& methodInfo, Protocol *protocol, SEL sel
+        auto argumentType = span([methodSignature getArgumentTypeAtIndex:i]);
 
-        if (*argumentType != '@') {
+        if (argumentType.front() != '@') {
```
I presume the old code was safe because of the null terminator on the string making it safe to read one past the end of the string. But can this hit `span` checking if the signature type is the empty string or is that impossible?

cdumez replied:
Yes, you have a point. I'm actually wondering if this is what's causing the API tests failures on iOS. I'm building right now to investigate.

cdumez replied:
I have confirmed this was the cause of the API tests failures on iOS. Fixing.

darinadler commented on Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectInterface.mm:
```diff
@@ -89,15 +87,15 @@ static void initializeMethod(MethodInfo& methodInfo, Protocol *protocol, SEL sel
         }
 
-        if (*(argumentType + 1) == '?') {
+        if (argumentType[1] == '?') {
```
I presume the old code was safe because of the null terminator on the string making it safe to read one past the end of the string. But can this hit `span` checking if the signature type is just the one character `@` or is that impossible?

darinadler commented on Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectInterface.mm:
```diff
@@ -140,22 +139,22 @@ static void initializeMethods(_WKRemoteObjectInterface *interface, Protocol *pro
 {
     unsigned conformingProtocolCount;
-    auto conformingProtocols = protocol_copyProtocolList(protocol, &conformingProtocolCount);
+    auto rawConformingProtocols = protocol_copyProtocolList(protocol, &conformingProtocolCount);
```
Same pattern here I’d think.

cdumez replied:
ditto.

darinadler commented on Source/WebKit/Shared/API/c/WKURLRequest.cpp:
```diff
@@ -61,13 +60,11 @@ WKStringRef WKURLRequestCopyHTTPMethod(WKURLRequestRef requestRef)
 {
     WebCore::ResourceRequest requestCopy(WebKit::toImpl(requestRef)->resourceRequest());
-    requestCopy.setHTTPBody(WebCore::FormData::create(std::span { WKDataGetBytes(body), WKDataGetSize(body) }));
+    requestCopy.setHTTPBody(WebCore::FormData::create(unsafeForgeSpan(WKDataGetBytes(body), WKDataGetSize(body))));
```
Seems like we want a helper function here that gets a `span` for a `WKData`.

cdumez replied:
Ok, adding one. Proposing `WKDataGetSpan()` in `WKData.h`

cdumez commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -1615,21 +1611,21 @@ def generate_message_names_implementation(receivers):
     result.append('namespace IPC::Detail {\n')
     result.append('\n')
-    result.append('const MessageDescription messageDescriptions[static_cast<size_t>(MessageName::Count) + 1] = {\n')
+    result.append('const std::array messageDescriptions {\n')
```
Hmm, this built on macOS but not glib ports unfortunately.

darinadler replied:
C++ standardization lets us down; no big deal I guess since it’s generated code. Maybe giving the array type a name in the header then is a nice small cleanup (I suggested that in a separate comment)? So we don’t need to write it out here a second time. Not important before landing this change, obviously. Just a tiny possible tweak.

cdumez replied:
>Maybe giving the array type a name in the header then is a nice small cleanup

This is what I did in the iteration that is currently up I believe.

cdumez committed:
Further reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebKit/
https://bugs.webkit.org/show_bug.cgi?id=282221

Reviewed by Geoffrey Garen and Darin Adler.

* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_names_header):
(generate_message_names_implementation):
* Source/WebKit/Shared/API/APIData.h:
(API::Data::create):
* Source/WebKit/Shared/API/Cocoa/WKRemoteObjectCoder.mm:
(-[WKRemoteObjectEncoder encodeBytes:length:forKey:]):
* Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectInterface.mm:
(initializeMethod):
(initializeMethods):
* Source/WebKit/Shared/API/c/WKArray.cpp:
(WKArrayCreate):
(WKArrayCreateAdoptingValues):
(WKArrayGetSize):
* Source/WebKit/Shared/API/c/WKData.cpp:
(WKDataCreate):
(WKDataGetSize):
* Source/WebKit/Shared/API/c/WKDictionary.cpp:
(WKDictionaryCreate):
(WKDictionaryCopyKeys):
* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringCreateWithUTF8CStringWithLength):
(WKStringGetUTF8CStringImpl):
(WKStringCopyJSString):
* Source/WebKit/Shared/API/c/WKURLRequest.cpp:
(WKURLRequestCopySettingHTTPBody):
(WKURLRequestSetDefaultTimeoutInterval):
* Source/WebKit/Shared/Cocoa/CoreIPCCFCharacterSet.h:
(WebKit::CoreIPCCFCharacterSet::dataReference const):
* Source/WebKit/Shared/Cocoa/CoreIPCDateComponents.mm:
* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtension::createHandlesForMachLookup):
* Source/WebKit/Shared/Cocoa/SandboxInitialiationParametersCocoa.mm:
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::tryApplyCachedSandbox):
* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
(WebKit::HistoryEntryDataDecoder::alignedBufferIsLargeEnoughToContain const):

Canonical link: https://commits.webkit.org/285917@main
